Refactor kbs client

Xynnn007 opened the pull request:
This PR mainly aims to tidy & refactor the code in `kbs_protocol`. Now KBS protocol supports both [passport](https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md#attestation-results-token-authentication) & [background check](https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md#http-cookie-authentication) model to access a resource. So the different ways are included in this PR

Also, some work aiming to tidy the `deps/crypto` was introduced.

To assert that the new `kbs_protocol` works, a unit test using a CoCo community published `kbs` image is impled.

see #292 

cc @jialez0 @mkulke

mkulke reviewed (commented):
I have only looked at the kbs client part so far, I will look at the passport part seperately:

I found that the `Attester` and the `EvidenceProvider` traits are very similar. Is there a reason to keep the Attester trait in the current form?

but if we still want to keep both traits we can create a "trait alias", it's a tiny bit of ceremony, but the `evidence_provider` code would be redundant, plz see [this commit](https://github.com/mkulke/guest-components/commit/314cb35ea7ef578465e4092b0a3ce6c60eea46ac) for reference

Xynnn007 commented on attestation-agent/attester/src/lib.rs:
```diff
@@ -22,38 +20,26 @@ pub mod sgx_dcap;
-    Sample,
-    Unknown,
-}
+pub struct BoxedAttester(pub Box<dyn Attester + Send + Sync>);
```
Great suggestion. Let me fix this

Xynnn007 commented on attestation-agent/attester/src/lib.rs:
```diff
@@ -22,38 +20,26 @@ pub mod sgx_dcap;
-    pub fn to_attester(&self) -> Result<Box<dyn Attester + Send + Sync>> {
-        match self {
-            Tee::Sample => Ok(Box::<sample::SampleAttester>::default()),
+    fn try_from(value: Tee) -> std::result::Result<Self, Self::Error> {
```
Great

Xynnn007 commented on attestation-agent/kbs_protocol/src/api.rs:
```diff
@@ -0,0 +1,20 @@
+
+use anyhow::Result;
+use async_trait::async_trait;
+use resource_uri::ResourceUri;
```
Nice suggestion. Added

Xynnn007 commented on attestation-agent/kbs_protocol/src/evidence_provider/native.rs:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    async fn get_tee_type(&self) -> Result<Tee> {
+        detect_tee_type()
```
What do you mean by this?

Xynnn007 commented:
> I found that the `Attester` and the `EvidenceProvider` traits are very similar. Is there a reason to keep the Attester trait in the current form?
> 
> but if we still want to keep both traits we can create a "trait alias", it's a tiny bit of ceremony, but the `evidence_provider` code would be redundant, plz see [this commit](https://github.com/mkulke/guest-components/commit/314cb35ea7ef578465e4092b0a3ce6c60eea46ac) for reference

Thanks for the reference! My first idea was that the EvidenceProvider and Attester are different:
1. Attester are native hardware driver/plugin
2. EvidenceProvider can be some client which will connect to an attester service to get evidence. This means the process uses `kbs_protocol` does not need to be the same process as the real attester process.

This design comes from `"a remote server(another process, or IMDS) will tell the concrete tee type"`. In type 2 we can impl such a client.

However they are actually the same thing. Your suggestion looks straightforward and better. Let me apply them.

mkulke commented:
> I discussed with @jialez0 and the detect_tee_type() should not be a trait function in Attester for we should firstly know which Tee is and then create an Attester object using the Tee enum.

yes, I agree. this makes sense. In this case `EvidenceProvider` and `Attester` probably have to be different traits, since they are doing things. I understand that `KbsClient._tee` is populated lazily in `background_check#rcar_handshake()` (and initialized with `None`).

Looking from the user side, a kbs client impl could look like this:

```rust
use kbs_protocol::{EvidenceProvider, KbsClientCapabilities};
use kbs_types::Tee;
use std::error::Error;

struct MyEvidenceProvider;

#[async_trait::async_trait]
impl EvidenceProvider for MyEvidenceProvider {
    async fn get_evidence(&self, _d: Vec<u8>) -> Result<String, Box<dyn Error>> {
        todo!();
    }
    async fn get_tee_type(&self) -> Result<Tee, Box<dyn Error>> {
        todo!();
    }
}

#[tokio::main]
async fn main() {
    let provider = Box::new(MyEvidenceProvider);
    let mut client =
        kbs_protocol::KbsClientBuilder::with_evidence_provider(provider, "https://kbs.com")
            .build()
            .unwrap();
    let resource_uri = "one/two/key".try_into().unwrap();
    let _ = client.get_resource(resource_uri).await.unwrap();
}
```

mkulke commented on attestation-agent/kbs_protocol/src/evidence_provider/mod.rs:
```diff
@@ -0,0 +1,23 @@
+#[async_trait]
+pub trait EvidenceProvider: Send + Sync {
+    /// Get evidence with as runtime data (report data, challege)
+    async fn get_evidence(&self, runtime_data: Vec<u8>) -> Result<String>;
```
We probably should not put anyhow types on a public trait. How about `Result<String, Box<dyn std::error::Error>>` or even a dedicated Error type?

mkulke commented on attestation-agent/kbs_protocol/src/evidence_provider/mod.rs:
```diff
@@ -0,0 +1,23 @@
+    async fn get_evidence(&self, runtime_data: Vec<u8>) -> Result<String>;
+
+    /// Get the underlying Tee type
+    async fn get_tee_type(&self) -> Result<Tee>;
```
See above comment about Result type

Xynnn007 replied:
nice idea. let me refactor this using thiserror tomorrow

fitzthum commented on attestation-agent/kbs_protocol/src/client/background_check.rs:
```diff
@@ -0,0 +1,270 @@
+            self.kbs_host_url, resource_uri.repository, resource_uri.r#type, resource_uri.tag
+        );
+
+        if self._tee.is_none() {
```
Maybe we should be checking `self.token` here instead? Notice that we make this exact same check as soon as we enter `rcar_handshake`.

jialez0 replied:
@Xynnn007 I agree, checking `self.token` here is better, and if `self.token.is_some()`, we needn't enter `rcar_handshake`.

fitzthum commented on attestation-agent/kbs_protocol/src/client/passport.rs:
```diff
@@ -0,0 +1,81 @@
+}
+
+#[async_trait]
+impl KbsClientCapabilities for KbsClient<Box<dyn TokenProvider>> {
```
I thought the point of the passport model was to return a Token, but we aren't implementing `PassportClientCapabilities` here at all. Am I missing something?

Xynnn007 replied:
see https://github.com/confidential-containers/guest-components/pull/304#discussion_r1281589440

fitzthum reviewed (commented):
Nice job. Some comments. Please bear with me as I attempt to understand all the changes.

Xynnn007 commented on attestation-agent/attester/src/lib.rs:
```diff
@@ -66,29 +52,30 @@ pub trait Attester {
     }
-    Tee::Unknown
+
+    bail!("Unknown tee type!");
```
Sounds good. `Option` is a good choice as actually there is not any error. Let me fix.

Xynnn007 commented on attestation-agent/kbs_protocol/src/token_provider/mock.rs:
```diff
@@ -0,0 +1,28 @@
+use super::TokenProvider;
+
+#[derive(Default)]
+pub struct MockedTokenProvider {}
```
Nice suggestion. @jialez0 suggests we extend this `TestTokenProvider` into a `ManualTokenProvider` which can also work as an example. At the same time we can `new()` it with a specified token.

jialez0 commented on attestation-agent/kbs_protocol/src/token_provider/mod.rs:
```diff
@@ -0,0 +1,66 @@
+    async fn get_token(&self) -> Result<(Token, TeeKeyPair)>;
+}
+
+#[derive(Clone, Debug)]
```
We should implement `ToString` or some thing like for this struct, so that we can get the original content of JWT.

Xynnn007 commented on attestation-agent/kbs_protocol/src/token_provider/mock.rs:
```diff
@@ -0,0 +1,28 @@
+#[derive(Default)]
+pub struct MockedTokenProvider {}
+
+/// A test Token expired in ~ 10years.
```
Emmm, actually I created this token using the code in KBS. To create it inside `kbs_protocol` crate would bring some more dependencies which is not expected.

My idea is that we can move `kbs_protocol` crate to `kbs` repo in future, as the former is the client of the latter.

After moving we can do this as they are in a same repo.

wdyt?

fitzthum replied:
Ok, that is fine.

Xynnn007 commented on attestation-agent/kbs_protocol/src/lib.rs:
````diff
@@ -3,312 +3,81 @@
+//! ```no_run
+//! use kbs_protocol::KbsClientBuilder;
+//! use kbs_protocol::evidence_provider::NativeEvidenceProvider;
+//! use kbs_protocol::{PassportClientCapabilities, KbsClientCapabilities};
````
yes, `PassportClientCapabilities` defines `get_token` ability. 

Also I find the names `passport` and `background check` are very confusing in the code. Let me rename them and just tell "how to perform a passport-model access or a background-check-model access in code"

fitzthum replied:
Ok, I was getting confused by those. I think it makes more sense now.

I notice that the passport feature doesn't have any attester associated with it. I assume we will be adding some new ones there?

Xynnn007 replied:
The passport feature Client does not need to care about attester. As attesters in KBS Client only helps to generate evidence in order to get token. The passport feature Client does not assume the origin of the token, and it just call `get_token` from TokenProvider.

Xynnn007 commented on attestation-agent/kbs_protocol/src/api.rs:
```diff
@@ -0,0 +1,20 @@
+
+#[async_trait]
+pub trait PassportClientCapabilities {
+    async fn get_token(&mut self) -> Result<(Token, TeeKeyPair)>;
```
Not related to the questions: let me delete the trait `PassportClientCapabilities` and just have a function `get_token()`, which is only to show that the client "can provide a token".

So the questions:

> So the token will always be linked to the TeeKeyPair and there is no option to provide any other user data? Will that cover all use cases?

Yes. This is a `KbsClient<Box<dyn EvidenceProvider>>` and it can generate a key pair and a token which certifies the token. RCAR  handshake (Remote attestation, evidence exchange ...) will be performed underlying

The key pair is bonded to a token. As KBS protocol defines the responsed data should use the key certified by the token to unwrap https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md#response, they should be given together

fitzthum replied:
I guess it's just following the protocol that we've settled on, so this should be fine. I think there might be some other possible use cases but we can worry about that later.

Xynnn007 commented on attestation-agent/kbs_protocol/src/api.rs:
```diff
@@ -0,0 +1,20 @@
+}
+
+#[async_trait]
+pub trait KbsClientCapabilities {
```
Ah, yes. Let me delete the trait `PassportClientCapabilities` which made me annoyed greatly.

Xynnn007 commented on attestation-agent/kbs_protocol/src/client/background_check.rs:
```diff
@@ -0,0 +1,270 @@
+        assert!(self.token.is_some());
+
+        let token = self.token.clone().unwrap();
+        let tee_key = self.tee_key.clone();
```
Do you mean by a long-term token provisioned by KBS?

There is some details. Now we support two different ways to access KBS:
1. Using the token in `BearAuth` http header
2. Using a same http session that pass a successful RCAR handshaking

Although the two mechanisms can be support altogether in `KbsClient<Box<dyn EvidenceProvider>>`, I choose to use way 2 to access the KBS. Thus the token retrieved is actually not used by the client when requesting for resources.

Also, the KBS can set the token expiration time when it starts [here](https://github.com/confidential-containers/kbs/blob/main/src/kbs/src/main.rs#L29) (Yes, the http session timeout and token expiration time now are the same!)

In this way a long-term token can be supported.

fitzthum replied:
Ok, so you're saying that if you want to use a the RCAR model and use the token for something, you can extend the timeout via the CLI to make it a long-lasting token. I think that is fine for now, but I think in the future the token given from the KBS in RCAR model should not be the same as the HTTP session. For one thing, I think that token should say something about the claims being made by the KBS, like which containers are approved to run in the guest.

Xynnn007 replied:
I agreed with you. Currently the implementation in KBS the two are the same, but we can extend this then.

mkulke commented on attestation-agent/kbs_protocol/src/client/mod.rs:
```diff
@@ -0,0 +1,40 @@
+/// This Client is used to connect to the remote KBS.
+pub struct KbsClient<T> {
+    /// TEE Type
+    pub(crate) _tee: Option<Tee>,
```
What does a (crate)-public `Option<Tee>` field intend to express?

An Option usually represents a value that may or may not exist in a struct. I suspect this is not case here and it's an implementation detail used it for lazy initialization.

If that's the case then we might want to make this field private, to prevent consuming code from messing with that in unintended ways. Maybe even use enum to document this intent:

```rust
enum ClientTee {
  Unitialized,
  Initializated(Tee),
}
```

fitzthum commented:
Ok, I think I've wrapped my head around this. Let me try to summarize. 

This PR gives us two methods to request a resource from a KBS. One is the RCAR method that we have been using, which is based on "evidence". The other uses a token instead of evidence. When we use a token, we skip over the typical challenge and attestation steps of RCAR by providing the token in the request and immediately getting a response.

Now to make things a bit more complicated, there is also a token involved in a normal RCAR request. The KBS sends a token to the client once attestation is successful. Here when we use the TokenProvider, we don't use this token from the KBS. Instead we use one generated by a TokenProvider. It doesn't seem like we've really implemented this yet, but presumably these tokens will come from the HW evidence of certain platforms.

Whether you have a platform that gives you "evidence" or a platform that gives you a token, you can connect to the KBS and get a resource. In both cases you can also get a token. In the RCAR case the token comes from the KBS. In the TokenProvider case, it comes directly from the TokenProvider.

What is the difference between a token and evidence? A token represents the attestation of the entire workload, whereas evidence is typically the initial state of the enclave. Hence the alternate terms passport and background check. A token is a passport and evidence is more like a background check. Again, you can get a passport/token either from RCAR or TokenProvider and you can connect to a KBS with RCAR or TokenProvider. With TokenProvider you can get a passport without connecting to the KBS.

Is this summary correct?

Xynnn007 commented:
@fitzthum Nice summury. About the `TokenProvider` I want to say more.

When I firstly designed this, two scenarios were considered:
- A KBS Client connects to a KBS using a token. The token is manually set by the user. Typically for testing, this is `TestTokenProvider`
- When we design the architecture of CDH and new AA, we decouple the "getting token" part and "using token to access resource" part. The "getting token" part should belong to AA, as it involves remote attestation. The "using token to access resource" should belong to CDH, as it cares about resource access. In this way, a programming problem is how to get a new token if the "using token to access resource" part finds the token is expired? Yes, call AA to generate a new one. 

The two scenarios are abstractly the same, s.t. we need a `TokenProvider` which is responsible for "generating new tokens". 
We will refactor AA later. At that time I'll add this `TokenProvider` which will call AA from CDH via ttRPC. 


We should also be careful. The trait `TokenProvider` makes it possible for a component inside Tee to get an token from somewhere else. Try to think about this: a malwared Tee gets a correct token to retrieve resource. So when we apply this crate, we must avoid insecure usages. This is beyond of the crate, but the overall design of CoCo/other platforms.

fitzthum reviewed (approved):
Ok, LGTM.

Looks like there is a merge conflict.

@mkulke can you take another look so we can merge and get the CDH stuff going

fitzthum commented:
> When we design the architecture of CDH and new AA, we decouple the "getting token" part and "using token to access resource" part. The "getting token" part should belong to AA, as it involves remote attestation. The "using token to access resource" should belong to CDH, as it cares about resource access. In this way, a programming problem is how to get a new token if the "using token to access resource" part finds the token is expired? Yes, call AA to generate a new one.

Hmmm. I guess I have misunderstood part of the design of the CDH. I had thought that the AA would not make any network calls. Instead it would just be responsible for getting the hardware measurement, which would be used by the KBS either as evidence or a token.

What you're saying is that we are going to use something more like the RATS passport model where the KBS will give the AA a token and then the CDH will use that token for future requests?

Xynnn007 commented:
> What you're saying is that we are going to use something more like the RATS passport model where the KBS will give the AA a token and then the CDH will use that token for future requests?

Yes

Xynnn007 commented:
> Looks like there is a merge conflict.

Rebased and fixed the conflict.

cc @mythi the conflict is about [the commit](https://github.com/confidential-containers/guest-components/pull/307/commits/96acbcf611658d5febc7b2a36dd035f0d06e7eed)

mythi commented:
> > Looks like there is a merge conflict.
> 
> Rebased and fixed the conflict.
> 
> cc @mythi the conflict is about [the commit](https://github.com/confidential-containers/guest-components/pull/307/commits/96acbcf611658d5febc7b2a36dd035f0d06e7eed)

yeah, looks like it was merged yesterday ahead of this so a rebase was needed. Thanks!

mkulke commented on attestation-agent/kbs_protocol/src/lib.rs:
````diff
@@ -3,316 +3,83 @@
+//! ```
+//!
+//! Note: everytime the token is found expired, the client will call the
+//! `evidence_provider` to retrieve a new token.
````
typo?

```suggestion
//! `token_provider` to retrieve a new token.
```

mkulke commented on attestation-agent/kbs_protocol/src/client/rcar_client.rs:
```diff
@@ -0,0 +1,298 @@
+        Ok((token, tee_key))
+    }
+
+    /// Perform RCAR handshake with the given kbs host. If succeeds, the [`KbsBackgoundCheckClient`] will
```
I think this is a leftover, the type doesn't exist anymore (and a typo, background is missing an r)

Xynnn007 replied:
Yes. fixed this

Xynnn007 commented:
@mkulke Thanks for the suggestions. I've fixed them. PTAL
